packages/http/src/router/index.ts

```diff
@@ -3,89 +3,94 @@ import { IHttpOperation, IServer } from '@stoplight/types';
 import { IMatch, MatchType } from './types';
 
 export const router: IRouter<IHttpOperation, IHttpRequest, IHttpConfig> = {
-  route: ({ resources, input }) => {
-    const matches = [];
+  route: ({ resources, input }): IHttpOperation => {
     const { path: requestPath, baseUrl: requestBaseUrl } = input.url;
-    const serverValidationEnabled = !!requestBaseUrl;
 
     if (!resources.length) {
-      throw NO_RESOURCE_PROVIDED_ERROR;
+      throw new Error(NO_RESOURCE_PROVIDED_ERROR);
     }
 
-    let anyMethodMatched = false;
-    let anyPathMatched = false;
-    let anyServerMatched = false;
-    let anyServerProvided = false;
+    const matches = resources.map<IMatch>(resource => {
+      const pathMatch = matchPath(requestPath, resource.path);
+      if (pathMatch === MatchType.NOMATCH)
+        return {
+          pathMatch,
+          methodMatch: MatchType.NOMATCH,
+          resource,
+        };
 
-    for (const resource of resources) {
-      if (!matchByMethod(input, resource)) continue;
-      anyMethodMatched = true;
+      const methodMatch = matchByMethod(input, resource) ? MatchType.CONCRETE : MatchType.NOMATCH;
 
-      const pathMatch = matchPath(requestPath, resource.path);
-      if (pathMatch !== MatchType.NOMATCH) anyPathMatched = true;
+      if (methodMatch === MatchType.NOMATCH) {
+        return {
+          pathMatch,
+          methodMatch,
+          resource,
+        };
+      }
 
       const { servers = [] } = resource;
-      let serverMatch: MatchType | null = null;
 
-      if (serverValidationEnabled) {
-        if (servers.length === 0) continue;
+      if (requestBaseUrl && servers.length > 0) {
+        const serverMatch = matchServer(servers, requestBaseUrl);
 
-        anyServerProvided = true;
-        serverMatch = matchServer(servers, requestBaseUrl as string);
-        if (serverMatch) anyServerMatched = true;
-      } else {
-        anyServerMatched = true;
-        anyServerProvided = true;
-      }
-
-      if (pathMatch !== MatchType.NOMATCH) {
-        matches.push({
+        return {
           pathMatch,
+          methodMatch,
           serverMatch,
           resource,
-        });
+        };
       }
-    }
 
-    if (!anyMethodMatched) {
-      throw NONE_METHOD_MATCHED_ERROR;
-    }
+      return {
+        pathMatch,
+        methodMatch,
+        serverMatch: null,
+        resource,
+      };
+    });
+
+    if (requestBaseUrl) {
+      if (matches.every(match => match.serverMatch === null)) {
+        throw new Error(NO_SERVER_CONFIGURATION_PROVIDED_ERROR);
+      }
 
-    if (!anyPathMatched) {
-      throw NONE_PATH_MATCHED_ERROR;
+      if (matches.every(match => !!match.serverMatch && match.serverMatch === MatchType.NOMATCH)) {
+        throw new Error(NO_SERVER_MATCHED_ERROR);
+      }
     }
 
-    if (!anyServerProvided) {
-      throw NO_SERVER_CONFIGURATION_PROVIDED_ERROR;
+    if (!matches.some(match => match.pathMatch !== MatchType.NOMATCH)) {
```

Comment:
I like your refactoring in general but whether it's more readable or not it's rather debatable ;)

Maybe you could increase the readability by extracting those one liners to small functions. For instance:

```suggestion
    if (noPathMatched(matches)) {
```

Reply:
Yeah, I think is a good idea! Will do!

Reply:
3b4b9cf